util/encoding.go

```diff
@@ -0,0 +1,243 @@
+package util
+
+import (
+	"encoding/base64"
+	"errors"
+	"unsafe"
+)
+
+var (
+	fibEncodeNumOutOfRangeErr = errors.New("the number to be encoded is out of range")
+	fibEncodeInvalidRange     = errors.New("the range is invalid")
+)
+
+func getByteSlice() []byte {
+	// Most string to be encoded are less than 8 bytes.
+	return make([]byte, 0, 8)
+}
+
+func NewBitStreamForWrite() *BitStream {
+	return NewBitStream(getByteSlice())
+}
+
+// enlarge the underlying byte slice.
+// In short, if the final size is less than the expThreshold, double the size of the underlying byte array,
+// otherwise, enlarge the size of the array by a specified step.
+func (bs *BitStream) enlarge(n uint16) {
+	final := int(bs.p+n+7) / 8
+	// Under most circumstances, the value of final will hit this if condition.
+	if final <= cap(bs.b) {
+		bs.b = bs.b[:final]
+		return
+	}
+
+	const expThreshold = 64
+	const step = 16
+	var temp []byte
+
+	if final <= expThreshold {
+		finalF := float32(final - 1)
+		// Find out the leftmost one's position.
+		pos := *(*int)(unsafe.Pointer(&finalF))>>23&255 - 127
+		size := 1
+		if pos >= 0 {
+			size = 1 << (pos + 1)
+		}
+		temp = make([]byte, final, size)
+	} else {
+		size := (final-expThreshold+step-1)/step*step + expThreshold
```

Comment:
Might want to add a parenthesis to make the order of operations clearer for humans.
`((final-expThreshold+step-1)/step)*step`
It took me a bit to parse this correctly in my head.